adapters/repos/db/vector/hnsw/index.go

```diff
@@ -744,3 +717,106 @@ func IsHNSWIndex(index any) bool {
 	_, ok := index.(*hnsw)
 	return ok
 }
+
+type nodeLevel struct {
+	nodeId uint64
+	level  int
+}
+
+func (h *hnsw) calculateUnreachablePoints() []uint64 {
+	h.RLock()
+	defer h.RUnlock()
+
+	visitedPairs := make(map[nodeLevel]bool)
+	candidateList := []nodeLevel{{h.entryPointID, h.currentMaximumLayer}}
+
+	for len(candidateList) > 0 {
+		currentNode := candidateList[len(candidateList)-1]
+		candidateList = candidateList[:len(candidateList)-1]
+		if !visitedPairs[currentNode] {
+			visitedPairs[currentNode] = true
+			node := h.nodes[currentNode.nodeId]
```

Comment:
I think there's another race here:

```suggestion
			h.shardedNodeLocks.RLock(currentNode.nodeId)
			node := h.nodes[currentNode.nodeId]
			node.Lock()
			neighbors := node.connectionsAtLowerLevelsNoLock(currentNode.level, visitedPairs)
			node.Unlock()
			h.shardedNodeLocks.RUnlock(currentNode.nodeId)
```

Basically:
- one lock for accessing the _slice_ slot/index
- one lock for accessing the node _content_